File: scripts/jarvis-ptt-autosend.py
#!/usr/bin/env python3
"""Jarvis PTT auto-send for Cursor.

Hold Ctrl+M to speak (Cursor hold-to-talk). On release, wait for STT to
finish filling the composer, then press Return so the agent runs.

Requires: System Settings → Privacy & Security → Accessibility → allow
Terminal / Cursor / python (whichever launches this script).

Stop: Ctrl+C, or kill the PID in ~/.cursor/jarvis-ptt-autosend.pid
"""
from __future__ import annotations

import os
import signal
import subprocess
import sys
import threading
import time

PID_FILE = os.path.expanduser("~/.cursor/jarvis-ptt-autosend.pid")
CURSOR_BUNDLE = "com.todesktop.230313mzl4w4u92"
# Delay after Ctrl+M release so Cursor can finish batch STT into the box.
SEND_DELAY_SEC = float(os.environ.get("JARVIS_PTT_SEND_DELAY", "2.0"))
KEYCODE_M = 46
KEYCODE_RETURN = 36


def frontmost_is_cursor() -> bool:
    try:
        out = subprocess.check_output(
            [
                "osascript",
                "-e",
                'tell application "System Events" to get bundle identifier of first process whose frontmost is true',
            ],
            text=True,
            stderr=subprocess.DEVNULL,
        ).strip()
        return out == CURSOR_BUNDLE
    except (subprocess.CalledProcessError, OSError):
        return False


def press_return() -> None:
    if not frontmost_is_cursor():
        return
    # key code 36 = Return
    subprocess.run(
        [
            "osascript",
            "-e",
            'tell application "System Events" to key code 36',
        ],
        check=False,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def write_pid() -> None:
    os.makedirs(os.path.dirname(PID_FILE), exist_ok=True)
    with open(PID_FILE, "w", encoding="utf-8") as f:
        f.write(str(os.getpid()))


def clear_pid() -> None:
    try:
        os.remove(PID_FILE)
    except OSError:
        pass


def main() -> int:
    try:
        from Quartz import (  # type: ignore
            CFMachPortCreateRunLoopSource,
            CFRunLoopAddSource,
            CFRunLoopGetCurrent,
            CFRunLoopRun,
            CGEventGetFlags,
            CGEventGetIntegerValueField,
            CGEventTapCreate,
            CGEventTapEnable,
            kCFRunLoopCommonModes,
            kCGEventFlagMaskControl,
            kCGEventKeyDown,
            kCGEventKeyUp,
            kCGHeadInsertEventTap,
            kCGKeyboardEventKeycode,
            kCGSessionEventTap,
        )
    except Exception as exc:  # noqa: BLE001
        print(
            "Need pyobjc Quartz (usually present on macOS Python). "
            f"Import failed: {exc}",
            file=sys.stderr,
        )
        return 1

    write_pid()
    pending: threading.Timer | None = None
    lock = threading.Lock()
    ctrl_m_down = False

    def cancel_pending() -> None:
        nonlocal pending
        with lock:
            if pending is not None:
                pending.cancel()
                pending = None

    def schedule_send() -> None:
        nonlocal pending

        def _fire() -> None:
            press_return()

        cancel_pending()
        with lock:
            pending = threading.Timer(SEND_DELAY_SEC, _fire)
            pending.daemon = True
            pending.start()

    def tap_callback(proxy, event_type, event, refcon):  # noqa: ANN001, ARG001
        nonlocal ctrl_m_down
        try:
            keycode = CGEventGetIntegerValueField(event, kCGKeyboardEventKeycode)
            flags = CGEventGetFlags(event)
            ctrl = bool(flags & kCGEventFlagMaskControl)
            if keycode != KEYCODE_M:
                return event
            if event_type == kCGEventKeyDown and ctrl:
                ctrl_m_down = True
                cancel_pending()
            elif event_type == kCGEventKeyUp and ctrl_m_down:
                ctrl_m_down = False
                if frontmost_is_cursor():
                    schedule_send()
        except Exception:  # noqa: BLE001
            pass
        return event

    tap = CGEventTapCreate(
        kCGSessionEventTap,
        kCGHeadInsertEventTap,
        0,
        (1 << kCGEventKeyDown) | (1 << kCGEventKeyUp),
        tap_callback,
        None,
    )
    if not tap:
        print(
            "Could not create event tap. Grant Accessibility to this process "
            "(System Settings → Privacy & Security → Accessibility), then rerun.",
            file=sys.stderr,
        )
        clear_pid()
        return 1

    CGEventTapEnable(tap, True)
    source = CFMachPortCreateRunLoopSource(None, tap, 0)
    CFRunLoopAddSource(CFRunLoopGetCurrent(), source, kCFRunLoopCommonModes)

    def _stop(*_args: object) -> None:
        cancel_pending()
        clear_pid()
        sys.exit(0)

    signal.signal(signal.SIGINT, _stop)
    signal.signal(signal.SIGTERM, _stop)

    print(
        f"Jarvis PTT auto-send running (delay {SEND_DELAY_SEC}s). "
        "Hold Ctrl+M in Cursor, speak, release — Return is sent automatically.",
        flush=True,
    )
    CFRunLoopRun()
    clear_pid()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/jarvis-ptt-autosend.swift
#!/usr/bin/env swift
import Cocoa
import Carbon.HIToolbox

/// Hold Ctrl+M in Cursor → on release, wait for STT, press Return.
/// Runs as Jarvis PTT.app (login item). Grant Accessibility once.
///
/// Cursor disables Voice Submit Keywords while hold-to-talk is held,
/// so this helper is required for hands-free send.

private let cursorBundle = "com.todesktop.230313mzl4w4u92"
private let sendDelay: TimeInterval = Double(ProcessInfo.processInfo.environment["JARVIS_PTT_SEND_DELAY"] ?? "2.2") ?? 2.2
private let pidFile = NSString(string: "~/.cursor/jarvis-ptt-autosend.pid").expandingTildeInPath
private let logFile = NSString(string: "~/.cursor/jarvis-ptt-autosend.log").expandingTildeInPath

private func log(_ msg: String) {
    let stamp = ISO8601DateFormatter().string(from: Date())
    let line = "[\(stamp)] \(msg)\n"
    if let data = line.data(using: .utf8) {
        if FileManager.default.fileExists(atPath: logFile) {
            if let handle = try? FileHandle(forWritingTo: URL(fileURLWithPath: logFile)) {
                handle.seekToEndOfFile()
                handle.write(data)
                try? handle.close()
            }
        } else {
            try? data.write(to: URL(fileURLWithPath: logFile))
        }
    }
    fputs(line, stderr)
}

private final class State {
    static let shared = State()
    var ctrlMDown = false
    var pending: DispatchWorkItem?
    var eventTap: CFMachPort?

    func frontmostIsCursor() -> Bool {
        NSWorkspace.shared.frontmostApplication?.bundleIdentifier == cursorBundle
    }

    func reenableTapIfNeeded() {
        guard let tap = eventTap else { return }
        CGEvent.tapEnable(tap: tap, enable: true)
    }

    func pressReturn() {
        guard frontmostIsCursor() else {
            log("skip Return — Cursor not frontmost")
            return
        }
        let src = CGEventSource(stateID: .hidSystemState)
        let down = CGEvent(keyboardEventSource: src, virtualKey: CGKeyCode(kVK_Return), keyDown: true)
        let up = CGEvent(keyboardEventSource: src, virtualKey: CGKeyCode(kVK_Return), keyDown: false)
        down?.post(tap: .cghidEventTap)
        up?.post(tap: .cghidEventTap)
        log("sent Return after Ctrl+M release")
    }

    func scheduleSend() {
        pending?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.pressReturn() }
        pending = work
        log("scheduled Return in \(sendDelay)s")
        DispatchQueue.main.asyncAfter(deadline: .now() + sendDelay, execute: work)
    }

    func cancelPending() {
        pending?.cancel()
        pending = nil
    }
}

private func writePid() {
    try? "\(ProcessInfo.processInfo.processIdentifier)".write(toFile: pidFile, atomically: true, encoding: .utf8)
}

private func clearPid() {
    try? FileManager.default.removeItem(atPath: pidFile)
}

writePid()
signal(SIGINT) { _ in
    State.shared.cancelPending()
    clearPid()
    exit(0)
}
signal(SIGTERM) { _ in
    State.shared.cancelPending()
    clearPid()
    exit(0)
}

let mask: CGEventMask = (1 << CGEventType.keyDown.rawValue) | (1 << CGEventType.keyUp.rawValue)
guard let tap = CGEvent.tapCreate(
    tap: .cgSessionEventTap,
    place: .headInsertEventTap,
    options: .defaultTap,
    eventsOfInterest: mask,
    callback: { _, type, event, _ in
        // Critical: macOS disables taps after idle timeout — re-enable or auto-Enter dies silently.
        if type == .tapDisabledByTimeout || type == .tapDisabledByUserInput {
            State.shared.reenableTapIfNeeded()
            return Unmanaged.passUnretained(event)
        }
        let keycode = event.getIntegerValueField(.keyboardEventKeycode)
        let flags = event.flags
        let ctrl = flags.contains(.maskControl)
        if keycode == 46 { // kVK_ANSI_M
            let state = State.shared
            if type == .keyDown && ctrl {
                state.ctrlMDown = true
                state.cancelPending()
            } else if type == .keyUp && state.ctrlMDown {
                state.ctrlMDown = false
                if state.frontmostIsCursor() {
                    state.scheduleSend()
                }
            }
        }
        return Unmanaged.passUnretained(event)
    },
    userInfo: nil
) else {
    log("ERROR: Could not create event tap — Accessibility not granted for Jarvis PTT.")
    clearPid()
    exit(1)
}

State.shared.eventTap = tap
let source = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, tap, 0)
CFRunLoopAddSource(CFRunLoopGetCurrent(), source, .commonModes)
CGEvent.tapEnable(tap: tap, enable: true)

// Keep the tap alive even if macOS disables it quietly (no callback in some cases).
Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { _ in
    State.shared.reenableTapIfNeeded()
}

log("OK: Jarvis PTT running (delay \(sendDelay)s). Hold Ctrl+M in Cursor → speak → release.")
CFRunLoopRun()
